### PR TITLE
Add support for unique events

### PR DESCRIPTION
**Pull request by imankulov**
This adds support for "ethernal" bitmapist events (a.k.a user properties)

### CODE REVIEW

File: bitmapist/__init__.py [hunk 1]
```diff
@@ -148,6 +151,8 @@ def mark_event(event_name, uuid, system='default', now=None, track_hourly=None,
         `datetime.utcnow()`
     :param :track_hourly Should hourly stats be tracked, defaults to
         bitmapist.TRACK_HOURLY
+    :param :track_hourly Should unique stats be tracked, defaults to
```
Comment (daremon) [inline bitmapist/__init__.py]: `:param: track_unique`

File: bitmapist/__init__.py [hunk 2]
```diff
@@ -100,6 +100,9 @@
 # of memory that Redis uses (especially with huge integers)
 TRACK_HOURLY = False
 
+# Should unique events be trackedi as default?
```
Comment (daremon) [inline bitmapist/__init__.py]: `tracked`i

File: bitmapist/__init__.py [hunk 3]
```diff
@@ -343,6 +352,31 @@ def __contains__(self, uuid):
...
+                   MixinEventsMisc, MixinBitOperations):
+
+    @classmethod
+    def from_date(cls, event_name, dt=None, system='default'):
+        return cls(event_name, system=system)
+
+    def __init__(self, event_name, system='default'):
+        self.event_name = event_name
+        self.system = system
+        self.redis_key = 'trackist_%s' % event_name
```
Comment (daremon) [inline bitmapist/__init__.py]: maybe use `_prefix_key(event_name, 'u')` ?

**Comment by imankulov** [comment]
@daremon thanks for the review.

> I'd really love the easiness of only writing `mark_event('active', unique=True)`

We do have enough arguments in `mark_event(...)`, and having both `unique` and `mark_unique` makes it very confusing. Maybe create a pair of functions `mark_unique_event()` / `unmark_unique_event()` instead?

**Comment by daremon** [comment]
> enough arguments

More than enough for sure.

> new functions `mark_unique_event()`

I'd prefer that over the 2 new class methods. Again personally I don't usually like "leaking" arguments to new functions, but seeing how these 2 sets of mark/unmark functions will just wrappers to `_mark()` it will be fine.

File: bitmapist/__init__.py [hunk 4]
```diff
@@ -201,6 +201,50 @@ def _mark(event_name, uuid, system='default', now=None,
...
+
+    :param :event_name The name of the event, could be "active" or "new_signups"
+    :param :uuid An unique id, typically user id. The id should not be huge,
+        read Redis documentation why (bitmaps)
+    :param :system The Redis system to use (string, Redis instance, or Pipeline
+
+    Examples::
+
+        # Mark id 42 as premium
+        mark_unique_event('premium', 42)
```
Comment (daremon) [inline bitmapist/__init__.py]: mark_unique

File: bitmapist/__init__.py [hunk 5]
```diff
@@ -201,6 +201,50 @@ def _mark(event_name, uuid, system='default', now=None,
...
+
+    :param :event_name The name of the event, could be "active" or "new_signups"
+    :param :uuid An unique id, typically user id. The id should not be huge,
+        read Redis documentation why (bitmaps)
+    :param :system The Redis system to use (string, Redis instance, or Pipeline
+
+    Examples::
+
+        # Mark id 42 as not premium anymore
+        unmark_unique_event('premium', 42)
```
Comment (daremon) [inline bitmapist/__init__.py]: unmark_unique

**Pull request merged**